react-components/src/architecture/concrete/observations/ObservationsTool.ts

```diff
@@ -0,0 +1,137 @@
+import { type TranslateKey } from '../../base/utilities/TranslateKey';
+import { ObservationsDomainObject } from './ObservationsDomainObject';
+import { BaseEditTool } from '../../base/commands/BaseEditTool';
+import { type VisualDomainObject } from '../../base/domainObjects/VisualDomainObject';
+import { type BaseCommand } from '../../base/commands/BaseCommand';
+import { CreateObservationCommand } from './CreateObservationCommand';
+import { SaveObservationsCommand } from './SaveObservationsCommand';
+import { DeleteObservationCommand } from './DeleteObservationCommand';
+import { createEmptyObservationProperties, isObservationIntersection } from './types';
+import { ObservationsView } from './ObservationsView';
+import { CustomObjectIntersectInput } from '@cognite/reveal';
+
+export class ObservationsTool extends BaseEditTool {
+  private _isCreating: boolean = false;
+
+  protected override canBeSelected(domainObject: VisualDomainObject): boolean {
+    return domainObject instanceof ObservationsDomainObject;
+  }
+
+  public get isCreating(): boolean {
+    return this._isCreating;
+  }
+
+  public setIsCreating(value: boolean): void {
+    this._isCreating = value;
+    if (value) {
+      this.renderTarget.setCrosshairCursor();
+    } else {
+      this.renderTarget.setNavigateCursor();
+    }
+  }
+
+  public override get icon(): IconType {
+    return 'Location';
+  }
+
+  public override get tooltip(): TranslateKey {
+    return { fallback: 'Show and edit observations' };
+  }
+
+  public override getToolbar(): Array<BaseCommand | undefined> {
+    return [
+      new CreateObservationCommand(),
+      new DeleteObservationCommand(),
+      new SaveObservationsCommand()
+    ];
+  }
+
+  public override onActivate(): void {
+    let domainObject = this.getObservationsDomainObject();
+    if (domainObject === undefined) {
+      domainObject = new ObservationsDomainObject(this.renderTarget.fdmSdk);
+      this.renderTarget.rootDomainObject.addChild(domainObject);
+    }
+    domainObject.setVisibleInteractive(true, this.renderTarget);
+  }
+
+  public override onDeactivate(): void {
+    const domainObject = this.getObservationsDomainObject();
+    domainObject?.setSelectedObservation(undefined);
+    domainObject?.setVisibleInteractive(false, this.renderTarget);
+  }
+
+  public override async onClick(event: PointerEvent): Promise<void> {
+    if (this._isCreating) {
+      await this.createPendingObservation(event);
+      return;
+    }
+    await this.selectOverlayFromClick(event);
+  }
+
+  public getObservationsDomainObject(): ObservationsDomainObject | undefined {
+    return this.rootDomainObject.getDescendantByType(ObservationsDomainObject);
+  }
+
+  public async save(): Promise<void> {
+    const domainObject = this.getObservationsDomainObject();
+    await domainObject?.save();
+  }
+
+  private async selectOverlayFromClick(event: PointerEvent): Promise<void> {
+    const camera = this.renderTarget.camera;
```

Comment:
It will not work, since it assume all other views as invisible.  Use     

```
  private async selectOverlayFromClick(event: PointerEvent): Promise<void> {
    const intersection = await this.getIntersection(event);
    if (intersection === undefined) {
      await super.onClick(event);
      return;
    }
    const domainObjectIntersection = intersection as DomainObjectIntersection;
    const domainObject = domainObjectIntersection?.domainObject;
    if (domainObject === undefined) {
      await super.onClick(event);
      return;
    }
    if (!(domainObject instanceof ObservationsDomainObject)) {
      await super.onClick(event);
      return;
    }
    const observation = domainObjectIntersection.userData as Observation;
    domainObject.setSelectedObservation(observation);
  }

```
This uses the architecture as it is meant to be used. The userData is Observation, to the overlay since the overlay is part of the view.

Reply:
I rewrite it to this:
```
private async selectOverlayFromClick(event: PointerEvent): Promise<void> {
    const intersection = await this.getIntersection(event);

    if (intersection === undefined || !isObservationIntersection(intersection)) {
      await super.onClick(event);
      return;
    }

    intersection.domainObject.setSelectedObservation(intersection.userData.getContent());
  }
```

where `isObservationIntersection` is defined in `types.ts`:

```
export const observationMarker = Symbol('observationSymbol');

export type ObservationIntersection = Omit<
  DomainObjectIntersection,
  'userData' | 'domainObject'
> & {
  marker: typeof observationMarker;
  domainObject: ObservationsDomainObject;
  userData: Overlay3D<Observation>;
};

export function isObservationIntersection(
  objectIntersection: AnyIntersection
): objectIntersection is ObservationIntersection {
  return (objectIntersection as ObservationIntersection).marker === observationMarker;
}
```
This lets us assume that the intersection we create has the right type for all fields, which simplifies the code a bit.